File: CSF.Screenplay.Selenium.JavaScriptWorkarounds/ScriptResources/UpdateSelectElementSelection.js
function executeScript(argsArray)
{
  'use strict';
  
  argumentsArrayValidator.validate(argsArray, 2, 3);

  function validateElement(element)
  {
    if(element === null || element === undefined)
      throw new Error('You must provide an HTML element object.');
    if(!(element instanceof HTMLSelectElement))
      throw new Error('The element must be an HTML <select> element.');
  }
  
  var optionUpdater = function()
  {
    function triggerChangeEvent(element)
    {
      var ev = document.createEvent('Event');
      ev.initEvent('change', true, true);
      element.dispatchEvent(ev);
    }

    function getOptionByIndex(selectElement, index)
    {
      return selectElement.item(index);
    }
    
    function getOptionByValue(selectElement, value)
    {
      var options = selectElement.options;
      for(var i = 0, len = options.length; i < len; i++)
        if(options[i].value === value) return options[i];
      return null;
    }
    
    function getOptionByText(selectElement, text)
    {
      var options = selectElement.options;
      for(var i = 0, len = options.length; i < len; i++)
        if(options[i].text === text) return options[i];
      return null;
    }
    
    function setSelectedByIndex(selectElement, index, selected)
    {
      var option = getOptionByIndex(selectElement, index);
      return setSelected(selectElement, option, selected);
    }
    
    function setSelectedByValue(selectElement, value, selected)
    {
      var option = getOptionByValue(selectElement, value);
      return setSelected(selectElement, option, selected);
    }
    
    function setSelectedByText(selectElement, text, selected)
    {
      var option = getOptionByText(selectElement, text);
      return setSelected(selectElement, option, selected);
    }
    
    function setSelected(selectElement, option, selected)
    {
      if(!option) return false;
      option.selected = selected;
      triggerChangeEvent(selectElement);
      return true;
    }
    
    function deselectAll(selectElement)
    {
      var options = selectElement.options;
      for(var i = 0, len = options.length; i < len; i++)
        options.item(i).selected = false;
      triggerChangeEvent(selectElement);
      return options.length > 0;
    }
    
    return {
      selectByIndex: function(ele, idx) { return setSelectedByIndex(ele, idx, true); },
      selectByValue: function(ele, val) { return setSelectedByValue(ele, val, true); },
      selectByText: function(ele, txt) { return setSelectedByText(ele, txt, true); },
      deselectByIndex: function(ele, idx) { return setSelectedByIndex(ele, idx, false); },
      deselectByValue: function(ele, val) { return setSelectedByValue(ele, val, false); },
      deselectByText: function(ele, txt) { return setSelectedByText(ele, txt, false); },
      deselectAll: deselectAll,
    };
    
  }();
  
  var
    element = argsArray[0],
    action = argsArray[1],
    value = argsArray[2];
  
  validateElement(element);
  
  if(!optionUpdater.hasOwnProperty(action))
    throw new Error("The action '" + action + "' was not recognised.");
  
  return optionUpdater[action](element, value);
}

File: CSF.Screenplay.Selenium.JavaScriptWorkarounds/ScriptResources/UpdateSelectElementSelection.cs
//
// UpdateSelectElementSelection.cs
//
// Author:
//       Craig Fowler <craig@csf-dev.com>
//
// Copyright (c) 2018 Craig Fowler
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
using System;
using CSF.Screenplay.Selenium.StoredScripts;

namespace CSF.Screenplay.Selenium.ScriptResources
{
  /// <summary>
  /// A stored JavaScript which sets the selected option(s) for an HTML <c>&lt;select&gt;</c> element.
  /// </summary>
  public class UpdateSelectElementSelection : ScriptResource
  {
    /// <summary>
    /// Gets the action name for deselecting every option.
    /// </summary>
    public static readonly string DeselectAllActionName = "deselectAll";

    /// <summary>
    /// Gets the action name for selecting an option by its zero-based index.
    /// </summary>
    public static readonly string SelectByIndexActionName = "selectByIndex";

    /// <summary>
    /// Gets the action name for selecting an option by its value.
    /// </summary>
    public static readonly string SelectByValueActionName = "selectByValue";

    /// <summary>
    /// Gets the action name for selecting an option by its displyed text.
    /// </summary>
    public static readonly string SelectByTextActionName = "selectByText";

    /// <summary>
    /// Gets the action name for deselecting an option by its zero-based index.
    /// </summary>
    public static readonly string DeselectByIndexActionName = "deselectByIndex";

    /// <summary>
    /// Gets the action name for deselecting an option by its value.
    /// </summary>
    public static readonly string DeselectByValueActionName = "deselectByValue";

    /// <summary>
    /// Gets the action name for deselecting an option by its displayed text.
    /// </summary>
    public static readonly string DeselectByTextActionName = "deselectByText";

    /// <summary>
    /// Gets the name of this script.
    /// </summary>
    /// <value>The name.</value>
    public override string Name => "a JavaScript to set the select options of an HTML <select> element";

    /// <summary>
    /// Gets a collection of scripts which the current script instance depends upon.
    /// </summary>
    /// <returns>The dependencies.</returns>
    protected override ScriptResource[] GetDependencies() => new [] { new ArgumentsArrayValidator() };
  }
}
